tests/modules/keyspace_events.c

```diff
@@ -96,17 +96,20 @@ static int KeySpace_NotificationModule(RedisModuleCtx *ctx, int type, const char
     return REDISMODULE_OK;
 }
 
-static int KeySpace_NotificationUnlink(RedisModuleCtx *ctx, int type, const char *event, RedisModuleString *key){
+static int KeySpace_NotificationRemoved(RedisModuleCtx *ctx, int type, const char *event, RedisModuleString *key){
     REDISMODULE_NOT_USED(ctx);
     REDISMODULE_NOT_USED(type);
 
-    if(strcmp(event, "unlink") == 0){
+    /* Open Key to check key exists */
+    RedisModuleKey *kp = RedisModule_OpenKey(ctx, key, REDISMODULE_READ);
```

Comment:
maybe improve the test by actually keeping the value of the key, and validating it in the tcl code?

Reply:
Do we need check all types or just string. It is trivial.

Reply:
I think just string is enough

Reply:
I found `dbUnshareStringValue` calls `dbOverwrite`. So when I use `RedisModule_StringDMA`, it will call `moduleNotifyKeyUnlink` last. I really afraid of this to cause more bugs.

Reply:
i think that in the case of of `dbUnshareStringValue`, it may have been wrong to call `dbOverwrite` with all it's side effects.
i.e. other places that call `dbOverwrite` and `setKey` are really logically replacing the existing key with a new one, but this call doesn't intend to do that.

i.e. `dbUnshareStringValue` should not call any module hooks, keyspace notifications, and so on (currently does call `moduleNotifyKeyUnlink`).

it currently does do `moduleNotifyKeyspaceEvent(NOTIFY_REMOVED)`, and possibly calls `mt->unlink`, but since this call is only done in case:
```
    serverAssert(o->type == OBJ_STRING);
    if (o->refcount != 1 || o->encoding != OBJ_ENCODING_RAW) {
```
then the `mt->unlink` call is unreachable, and i think the `NOTIFY_REMOVED` is wrong.

@guybe7 @MeirShpilraien WDYT?

i think instead of calling `dbOverwrite`, `dbUnshareStringValue` should manually do a small sub-set of it's tasks.

Reply:
@oranagra I agree, it does not really override the data nor deleting anything, it's an internal implementation detail.